add support for oci-layout build-context

deitch opened the pull request:
Adds to buildx the oci-layout: build-context capabilities added to buildkit in [buildkit 2827](http://github.com/moby/buildkit/pull/2827).

Recommended by @tonistiigi . Remains incomplete, some things I wasn't sure about. Will comment inline.

deitch commented:
Pulling in the updated buildkit lib updated a whole bunch of other things. go mod worked its magic. It all worked and compiled, so I didn't complain.

deitch commented on build/build.go:
```diff
@@ -1235,6 +1238,33 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+			target.OCIStores = map[string]content.Store{
+				k: store,
+			}
+			target.FrontendAttrs["context:"+k] = "oci-layout:" + k
```
This didn't seem to make sense to me, but I wasn't sure what else we needed to put here, if anything? We need to have the name of the context, and a unique store ID. The name gets mapped to the store in `target.OCIStores` (2 lines above), and then the context to the unique store ID here. So should this be like I did here?

deitch commented on build/build.go:
```diff
@@ -1235,6 +1238,33 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+				return nil, errors.Wrapf(err, "invalid oci-layout context %s, must be oci-layout:///path/to/layout@sha256:hash", v.Path)
+			}
+			// check that we had a digest
+			digested, ok := named.(reference.Digested)
```
I wasn't sure where this goes. In the client test in buildkit [here](https://github.com/moby/buildkit/blob/a6a114a1a476c99c2501aa34811159b849df4005/client/client_test.go#L1669-L1689), we pulled the digest (which I can get from `digested.Digest()`) and put it in `llb.OCILayout(storeID, digest)`, which is used in the `client.Solve(ctx, def, opts)` in `def`.

Where do we put the digest here in buildx so it knows that the build-context `foo` refers not only to the OCI store, but to a specific digest?

tonistiigi commented on build/build.go:
```diff
@@ -1235,6 +1238,33 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+				return nil, errors.Wrapf(err, "invalid store at %s", localPath)
+			}
+			// now we can add it
+			target.OCIStores = map[string]content.Store{
```
This needs to add a store, not replace the whole map

deitch replied:
Good point. It should check if the map is nil, if so create it, and then add it either way.

tonistiigi commented on build/build.go:
```diff
@@ -1235,6 +1238,33 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+		if strings.HasPrefix(v.Path, "oci-layout://") {
+			pathAlone := strings.TrimPrefix(v.Path, "oci-layout://")
+			// need to add "dummy.io", because reference.Parse*() funcs require we do not start with a '/'
+			named, err := reference.ParseNormalizedNamed(fmt.Sprintf("dummy.io%s", pathAlone))
```
Rather than this hack just split by `@` and then `digest.Parse` maybe.

deitch replied:
One ugly hack vs another? I was _so_ disappointed than `reference.Parse` returns a thing that does not have component parts that can be looked up, like `url.Parse()` does.

tonistiigi commented on build/build.go:
```diff
@@ -1235,6 +1238,33 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+			target.OCIStores = map[string]content.Store{
+				k: store,
+			}
+			target.FrontendAttrs["context:"+k] = "oci-layout:" + k
```
The value in here nees to include digest that you are expecting in https://github.com/moby/buildkit/blob/master/frontend/dockerfile/builder/build.go#L892

deitch replied:
Ah, that is it. I went through our buildkit PR, and could not recall which part did this. That helps.

deitch replied:
Take a look, I _think_ I got it right.

tonistiigi commented on docs/reference/buildx_build.md:
````diff
@@ -126,6 +126,30 @@ FROM alpine
+Source an image from a local [OCI layout compliant directory](https://github.com/opencontainers/image-spec/blob/main/image-layout.md):
+
+```console
+$ docker buildx build --build-context foo=oci-layout:///path/to/local/layout@sha256:abcd12345 .
````
I think we should allow this input without the digest as well in here. If no digest is set it can look up `index.json` inside this dir and see if it has one descriptor and then set it as the expected digest. If there is ambiguity then it can produce an error.

deitch replied:
I am hesitant about this, as it wades into a deep unresolved issue around what actually should be in `index.json`. I have seen those that _only_ include tags, others that include entire image refs, making the layout directory essentially into a full registry. I have implemented. both ways.

containerd does a bit of a hybrid: directory _is_ a full registry, but tag->manifest-hash is in the boltdb.

Let's get this working with the hash required, as is, and then let's deal with supporting hash-less - both in buildkit and in buildx - as a separate iteration.

tonistiigi reviewed (commented):
In addition

deitch commented:
Most things resolved, one open to be sure that this is right, another subject to discussion.

deitch commented:
Fixed the go mod issues. Phew. 

I am seeing vet issues with:

```
# github.com/docker/buildx/bake
bake/bake.go:885:69: cannot use os.Stderr (variable of type *os.File) as type *configfile.ConfigFile in argument to authprovider.NewDockerAuthProvider
# github.com/docker/buildx/bake
vet: bake/bake.go:885:69: cannot use os.Stderr (variable of type *os.File) as *configfile.ConfigFile value in argument to authprovider.NewDockerAuthProvider
```

I didn't think I changed those, but perhaps downstream dependencies?

deitch commented:
Yeah, it looks like the definition of `github.com/moby/buildkit/session/auth/authprovider.NewDockerAuthProvider()` changed in buildkit from:

```go
func NewDockerAuthProvider(stderr io.Writer) session.Attachable {
	return &authProvider{
		config:      config.LoadDefaultConfigFile(stderr),
		seeds:       &tokenSeeds{dir: config.Dir()},
		loggerCache: map[string]struct{}{},
	}
}
```

to

```go
func NewDockerAuthProvider(cfg *configfile.ConfigFile) session.Attachable {
	return &authProvider{
		authConfigCache: map[string]*types.AuthConfig{},
		config:          cfg,
		seeds:           &tokenSeeds{dir: config.Dir()},
		loggerCache:     map[string]struct{}{},
	}
}
```

I got it.

tonistiigi commented on build/build.go:
```diff
@@ -1235,6 +1238,32 @@ func LoadInputs(ctx context.Context, d driver.Driver, inp Inputs, pw progress.Wr
+			}
+			target.OCIStores[k] = store
+
+			target.FrontendAttrs["context:"+k] = fmt.Sprintf("oci-layout:%s@sha256:%s", k, hash.Encoded())
```
Did you forget to push? Looks the same as before.

deitch replied:
Doh!

deitch replied:
Done

deitch commented:
Figured out the failing CI test, updated, should be good now.

deitch commented:
How done stand here @tonistiigi ?

deitch commented:
Been almost a month. I just rebased.

Is there anything else we need to get this merged in?

deitch commented on docs/reference/buildx_build.md:
````diff
@@ -126,6 +126,30 @@ FROM alpine
 COPY --from=project myfile /
 ```
 
+Source an image from a local [OCI layout compliant directory](https://github.com/opencontainers/image-spec/blob/main/image-layout.md):
````
Done. Back to you @tonistiigi
